lib/redis_memo/memoize_method.rb

```diff
@@ -81,14 +81,63 @@ def self.extract_dependencies(ref, *method_args, &depends_on)
 
   def self.get_or_extract_dependencies(ref, *method_args, &depends_on)
     if RedisMemo::Cache.local_dependency_cache
-      RedisMemo::Cache.local_dependency_cache[ref] ||= {}
-      RedisMemo::Cache.local_dependency_cache[ref][depends_on] ||= {}
-      RedisMemo::Cache.local_dependency_cache[ref][depends_on][method_args] ||= extract_dependencies(ref, *method_args, &depends_on)
+      RedisMemo::Cache.local_dependency_cache[ref.class] ||= {}
+      RedisMemo::Cache.local_dependency_cache[ref.class][depends_on] ||= {}
+      named_args = exclude_anonymous_args(depends_on, ref, method_args)
+      RedisMemo::Cache.local_dependency_cache[ref.class][depends_on][named_args] ||= extract_dependencies(ref, *method_args, &depends_on)
     else
       extract_dependencies(ref, *method_args, &depends_on)
     end
   end
 
+  # We only look at named method parameters in the dependency block in order to define its dependent
+  # memos and ignore anonymous parameters, following the convention that nil or :_ is an anonymous parameter.
+  # Example:
+  # ```
+  #    def method(param1, param2)
+  #    end
+  #
+  #    memoize_method :method do |_, _, param2|`
+  #      depends_on RedisMemo::Memoizable.new(param2: param2)
+  #    end
+  # ```
+  #  `exclude_anonymous_args(depends_on, ref, [1, 2])` returns [2]
+  def self.exclude_anonymous_args(depends_on, ref, args)
+    return [] if depends_on.parameters.empty? or args.empty?
+
+    positional_args = []
+    kwargs = {}
+    named_rest = false
+    depends_on_args = [ref] + args
+    options = depends_on_args.extract_options!&.clone
+
+    depends_on.parameters.each_with_index do |param, i|
+      unless param.size != 2 || param.last == :_
+        # Defined by https://github.com/ruby/ruby/blob/22b8ddfd1049c3fd1e368684c4fd03bceb041b3a/proc.c#L3048-L3059
+        case param.first
+        when :opt, :req
+          positional_args << depends_on_args[i]
+        when :rest
+          named_rest = true
+          positional_args.concat(depends_on_args[i..-1])
+        when :key, :keyreq
+          kwargs[param.last] = options.delete(param.last)
+        when :keyrest
+          kwargs.merge!(options)
+        else
+          raise(RedisMemo::ArgumentError, "#{param.first} argument isn't supported in the dependency block")
+        end
+      end
+    end
+    if !kwargs.empty?
+      positional_args + [kwargs]
+    elsif named_rest && !options.empty?
+        positional_args + [options]
```

Comment:
Why would we have this special case? `kwargs` should always be used right? when it's empty, it means  `keyrest` is not named.

Reply:
If the rest option was the last parameter, then we want it to capture any options arguments. Otherwise,

```ruby
memoize_method do :test |_, *args|
test(1, 1, a: 2, b: 3)
```
Would miss `a: 2, b: 3`

Reply:
oooh lol tricky tricky

Reply:
Actually, maybe we should still fill `kwargs` for this case
```
test(1, 1, a: 2, b: 3)
```